crates/wit-parser/src/ast.rs

```diff
@@ -1,24 +1,329 @@
+                        // },
+                        ExternKind::Id(id) => {
+                            let interface = match interface_map.get(&*id.name) {
+                                Some(interface) => interface.clone(),
+                                None => {
+                                    return Err(Error {
+                                        span: id.span,
+                                        msg: format!("{} not defined in document", id.name),
+                                    }.into());
+                                },
+                            };
+                            if world.exports.insert(name.to_string(), interface).is_some() {
+                                return Err(Error {
+                                    span: span,
+                                    msg: format!("duplicate export name {}", name),
+                                }
+                                .into());
+                            }
+                        },
+                    }
+                }
+            },
+            // nothing to do if no worlds were defined
+            None => {},
+        }
+
+        Ok(world)
+    }
+}
+
+pub enum DocumentItem<'a> {
+    Interface(Interface<'a>),
+    World(World<'a>),
+}
+
+impl<'a> DocumentItem<'a> {
+    fn parse(tokens: &mut Tokenizer<'a>, docs: Docs<'a>) -> Result<DocumentItem<'a>> {
+        match tokens.clone().next()? {
+            Some((_span, Token::Interface)) => Interface::parse(tokens, docs).map(DocumentItem::Interface),
+            Some((_span, Token::World)) => World::parse(tokens, docs).map(DocumentItem::World),
+            other => Err(err_expected(tokens, "`default`, `world` or `interface`", other).into()),
+        }
+    }
+}
+
+pub struct World<'a> {
+    docs: Docs<'a>,
+    name: Id<'a>,
+    items: Vec<WorldItem<'a>>,
+}
+
+impl<'a> World<'a> {
+    fn parse(tokens: &mut Tokenizer<'a>, docs: Docs<'a>) -> Result<Self> {
+        tokens.expect(Token::World)?;
+        let name = parse_id(tokens)?;
+        let items = Self::parse_items(tokens)?;
+        Ok(World { docs, name, items })
+    }
+
+    fn parse_items(tokens: &mut Tokenizer<'a>) -> Result<Vec<WorldItem<'a>>> {
+        tokens.expect(Token::LeftBrace)?;
+        let mut items = Vec::new();
+        loop {
+            let docs = parse_docs(tokens)?;
+            if tokens.eat(Token::RightBrace)? {
+                break;
+            }
+            items.push(WorldItem::parse(tokens, docs)?);
+        }
+        Ok(items)
+    }
+
+    fn imports(&self) -> impl Iterator<Item = &Import<'a>> {
+        self.items.iter().filter_map(|item| match item {
+            WorldItem::Import(import) => Some(import),
+            _ => None,
+        })
+    }
+
+    fn exports(&self) -> impl Iterator<Item = &Export<'a>> {
+        self.items.iter().filter_map(|item| match item {
+            WorldItem::Export(export) => Some(export),
+            _ => None,
+        })
+    }
```

Comment:
If possible I think it might be best to remove this style of iterators because it makes it harder to add more cases in the future and automatically have the compiler inform us what needs updating.